File: dazuma/euler/problem_032.ex
defmodule Euler032 do

  def permutation(_list, 0), do: [[]]
  def permutation(list, num), do:
    (for elem <- list, rest <- permutation(list -- [elem], num - 1), do: [elem | rest])

  def main() do
    import List, only: [to_integer: 1]
    import Enum
    digits = '123456789'
    permutation(digits, 4)
      |> map(fn(front) ->
        product = to_integer(front)
        permutation(digits -- front, 5) |> find_value(0, fn (back) ->
          if (back |> take(1) |> to_integer) * (back |> drop(1) |> to_integer) == product or
             (back |> take(2) |> to_integer) * (back |> drop(2) |> to_integer) == product,
             do: product, else: false
        end)
      end)
      |> sum
  end

end

Euler032.main |> IO.puts

# Runtime: 0.58s

File: dazuma/euler/problem_032.rb
module Euler032

  def self.main
    digits = [1,2,3,4,5,6,7,8,9]
    digits.permutation(4)
      .map{ |front|
        product = front.join.to_i
        (digits - front).permutation(5).any?{ |back|
          back[0..0].join.to_i * back[1..-1].join.to_i == product ||
            back[0..1].join.to_i * back[2..-1].join.to_i == product
        } ? product : 0
      }
      .reduce(:+)
  end

end

puts Euler032.main

# Runtime: 1.60s
